samples/validate-utility-network-topology/src/main/java/com/esri/arcgismaps/sample/validateutilitynetworktopology/components/ValidateUtilityNetworkTopologyViewModel.kt

```diff
@@ -0,0 +1,664 @@
+
+    // Graphic to represent the starting point of the downstream trace
+    private val startingLocationGraphic = Graphic(
+        symbol = SimpleMarkerSymbol(
+            style = SimpleMarkerSymbolStyle.Cross, color = Color.green, size = 25f
+        )
+    )
+
+    // Overlay to show the starting location for tracing
+    val graphicsOverlay = GraphicsOverlay(listOf(startingLocationGraphic))
+
+    // The utility network used for tracing
+    private val utilityNetwork: UtilityNetwork
+        get() = arcGISMap.utilityNetworks.first()
+
+    // ServiceGeodatabase from the utility network, used for switching to a new version
+    private var serviceGeodatabase: ServiceGeodatabase? = null
+
+    // Parameters for a downstream trace
+    private var traceParameters: UtilityTraceParameters? = null
+
+    // Track the state of utility network capabilities
+    private val _canGetState = MutableStateFlow(false)
+    val canGetState = _canGetState.asStateFlow()
+
+    // Track the state of utility network topology
+    private val _canTrace = MutableStateFlow(false)
+    val canTrace = _canTrace.asStateFlow()
+
+    // Track the state when we have dirty areas or errors
+    private val _canValidateNetworkTopology = MutableStateFlow(false)
+    val canValidateNetworkTopology = _canValidateNetworkTopology.asStateFlow()
+
+    // Track the state when selections can be cleared from map
+    private val _canClearSelection = MutableStateFlow(false)
+    val canClearSelection = _canClearSelection.asStateFlow()
+
+    // A collapsable status message to display in the UI
+    private val _statusMessage = MutableStateFlow("")
+    val statusMessage = _statusMessage.asStateFlow()
+
+    // Currently selected feature that the user is editing
+    private val _selectedFeature = MutableStateFlow<ArcGISFeature?>(null)
+    val selectedFeature = _selectedFeature.asStateFlow()
+
+    // Field being edited on the selected feature (e.g. "devicestatus" or "nominalvoltage")
+    private var selectedField: Field? = null
+
+    // List of coded values from the above field's domain
+    private val _fieldValueOptions = MutableStateFlow<List<CodedValue>>(emptyList())
+    val fieldValueOptions = _fieldValueOptions.asStateFlow()
+
+    // Selected field coded value from the domain
+    private val _selectedFieldValue = MutableStateFlow<CodedValue?>(null)
+    val selectedFieldValue = _selectedFieldValue.asStateFlow()
+
+    // Keep track of the current visible area to validate network topology
+    private val _currentVisibleArea = MutableStateFlow(Envelope(Point(0.0, 0.0), Point(0.0, 0.0)))
+
+    /**
+     * Add credentials and loads the map & utility network. Then creates & switches to a
+     * new version for editing. Sets up the default trace parameters and adds the dirty area
+     * table as a feature layer to the map. Sets the button states based on
+     * utility network capabilities.
+     */
+    suspend fun initialize() {
```

Comment:
could this be done inside a `init {` block?
i see that you're tracking the view model initialization with a flag on `ValidateUtilityNetworkTopologyScreen`. You could move the flag here and track its state like on `ValidateUtilityNetworkTopologyScreen`

Reply:
@eri9000 I followed the feedback on a previous [PR comment](https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/pull/300#discussion_r1913747724) from the Trace utility network sample:  

> It is preferable to not launch a coroutine from the constructor. Is it possible to simply move this code into a suspend function called something like initialize(), and call it from a launched coroutine in onCreate()?

Leaving this as is, if that's alright!